network/ieee802_11_prism.ksy

```diff
@@ -0,0 +1,113 @@
+meta:
+  id: ieee802_11_prism
+  title: prism II headers for 802.11 capture
+  license: Unlicense
+  endian: le
+  encoding: ascii
+  xref:
+    wikidata: Q7245672
+doc: |
+  Prism II headers are usually used by wlan cards based on Intersil / Conexant [Prism II](https://en.wikipedia.org/wiki/Prism_(chipset)).
+
+doc-ref: https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html
+
+seq:
+  - id: message_code
+    type: u4
+  - id: size
+    type: u4
+  - id: device_name
+    type: strz
+    size: 16
+  - id: fields
+    type: fields
+    size: fields_size
+  - id: body
+    size-eos: true
+instances:
+  header_size:
+    value: 4 + 4 + 16
+  fields_size:
+    value: size - header_size
+types:
+  fields:
+    seq:
+      - id: fields
+        type: field
+        repeat: eos
+    types:
+      field:
+        seq:
+          - id: header
+            type: header
+          - id: is_not_supplied
+            type: u2
+          - id: size
+            type: u2
+          - id: value
+            size: size
+            type:
+              switch-on: header.type
+              cases:
+                'type::host_time': u4
+                'type::mac_time': u4
+                'type::channel': u4
+                'type::rssi': u4
+                'type::signal_quality': u4
+                'type::noise': u4
+                'type::data_rate': u4
+                'type::is_tx': u4
+                'type::frame_length': u4
+        instances:
+          is_supplied:
+            value: is_not_supplied == 0
+        types:
+          header:
+            # all of these are wild guesses
+            seq:
+              - id: header_size
+                type: b4
+              - id: mandatory_part_offset
+                type: b4
+              - id: placeholder
+                type: placeholder
+                size: header_size - 1
+            instances:
+              mandatory_part_offset_from_first_byte:
+                value: mandatory_part_offset - 1
+              mandatory_part:
+                pos: mandatory_part_offset_from_first_byte / 2
+                type: mandatory_part(mandatory_part_skip_nibble)
+                io: placeholder._io
+              mandatory_part_skip_nibble:
+                value: mandatory_part_offset_from_first_byte % 2 == 1
+              type:
+                value: mandatory_part.type
```

Comment:
Oh my God, this is so much over-engineered! Let me remind you that according to the [specification](https://www.tcpdump.org/linktypes/LINKTYPE_IEEE802_11_PRISM.html) from `doc-ref` (yeah, I know it's bad to rely on a single source, but I don't know any others), the first byte of the DID field type (here split into `header_size` and `mandatory_part_offset` fields) is either **0x41** or **0x44**, nothing more. So your spec permits lots of combinations that will be rejected by any other parser implementation, and that won't ever generated by any reputable serialization implementation. OK, this is not so serious yet. What is worse, if we imagine that this spec may be used for serialization in the future, it would be very easy to start generating invalid packets that don't conform to the specification and aren't recognized by anything. Would it be the programmer's fault that he set some clearly incorrect values? Nope, he just looked into the KSY and saw that `header_size` and `mandatory_part_offset` are just plain `b4` bit-integers, so apparently all values are permitted, right? Wrong.

Of course, adding `valid/eq: 4` to `header_size` and `valid/any-of: [1, 4]` to `mandatory_part_offset` would address this issues, but then the `mandatory_part_*` stuff seems unnecessarily complex and cumbersome, when you have just 2 options to handle.

Please don't devise this **artificial structure** just for fun to make things more complex and unclear. Instead, stick to the specification and make simple things in a simple way. When the specification talks about two possible options, don't try to map random bits to their _potential_ meaning (as it currently fits) in order to invent a general formula or rule that magically handles the full theoretical range of inputs, but rather create an enum with these two permitted values and choose the behavior accordingly. The `.ksy` specification should match the format it describes as closely as possible, not try to outsmart the format authors themselves with some _wild guesses_. It helps not to think of the `.ksy` spec as a tool for generating parsing code, but rather as a format documentation by itself.

Reply:
> Oh my

Thou shalt not take the name in vain.

>according to the specification
>stick to the specification 

**There is no spec**, all the "specs" are reverse-engineered, so they are just wild guesses matching the files seen by the authors. We can say that all the reverse-engineered "specs" are overfitted. Mine is no better, but it just incorporates some domain knowledge about file formats in general. I mean, if we assumme that in all the cases the same file format is used, and we see that the offset varies, it is usually 3 cases: 
* the offset is encoded in the file itself explicitly as a number, relative to a fixed feature
* the offset is fixed, but relative to a feature which offset varies
* the offset is encoded, and the position of the feature is encoded

So, what else we know about this format?
* After this header another stuff starts untill the end.
* in the header itself there is pretty little stuff to be used as a base forz a fixed offset

So, I guesed it is likely that the offset is encoded, and, since sometimes the stuff starts from the second byte, the offset is likely to be encoded in the first one.

>So your spec permits lots of combinations that will be rejected by any other parser implementation

Yes. And it is OK. We just don't know, am I right, or are the "specs" writers are right, or none of us is right.

>What is worse, if we imagine that this spec may be used for serialization in the future, it would be very easy to start generating invalid packets

We don't know if they are valid or not. And what is good, if any official tools parsing these files can be ever found, we can be able to check, if I guessed right or not.